content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+```
+
+The option -o indicates the path and file name for the tracing data, so any path and name can be used. In this example we can use an arbitrary folder called “tracing” just outside Zephyr's main folder.
+
+After some seconds of acquisition we can hit CTRL+C to stop recording and get back to the shell in the host PC.
+
+The tracing format CTF requires data typing to be decoded, so the metadata (tracing format description) file needs to be present in the same folder:
+
+```shell
+cp subsys/tracing/ctf/tsdl/metadata ../tracing/
+```
+
+Finally, to parse the CTF data and show the actual trace:
+
+```shell
+babeltrace2 ../tracing/
+```
+
+The output is something like:
+
+<a href="img/img001.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img001.webp"
+        alt=""
+        caption="Sample trace output"
+    >}}
+</a>
+
+This log shows the execution order and timing of the traced objects, allowing us to visualize how the software is executing in the acquired interval. This gives the possibility of running statistics on how the CPU resources are used and easily spot functions that are taking too long to execute, or are being executed at wrong moments, or more times than expected.
+
+### Objects monitored
+
+It is possible to control which kinds of objects are monitored. The following are currently supported [[1]]:
+
+```c
+struct k_timer *_track_list_k_timer;
+struct k_mem_slab *_track_list_k_mem_slab;
+struct k_sem *_track_list_k_sem;
+struct k_mutex *_track_list_k_mutex;
+struct k_stack *_track_list_k_stack;
+struct k_msgq *_track_list_k_msgq;
+struct k_mbox *_track_list_k_mbox;
+struct k_pipe *_track_list_k_pipe;
+struct k_queue *_track_list_k_queue;
+struct k_event *_track_list_k_event;
+```
+
+Kconfig's menu *Tracing Configuration* will allow to select which objects will be traced, with options like shown below:
+
+```text
+TRACING_SYSCALL=y/n
+TRACING_ISR=y/n
+TRACING_SEMAPHORE=y/n
+TRACING_HEAP=y/n
+TRACING_STACK=y/n
+TRACING_PM=y/n
+(...)
+```
+
+If a large amount of data is to be generated, this might be useful to limit tracing only to the objects of interest, both to save buffer and provide a cleaner data acquisition.
+
+Another interesting feature is the possibility of defining user functions to be called at tracing events, as *headers* shown below. In a debugging situation or when external equipment (e.g oscilloscope or data logger) is used, it can be useful to add custom code to the callbacks to drive a GPIO, for instance, in order to sync external acquisition data with the log provided by the *tracing* subsystem.
+
+<a href="img/img002.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img002.webp"
+        alt=""
+        caption="Trace user functions"
+    >}}
+</a>
+
+The following macros [[2]] are placed on Zephyr's subsystems code to track important functions of the operating system:

```

Comment:
I see that the intention here is to show the number 2 that refers to link 2 in additional links. However, the rendered version of the section _Additional links_ does not have numbers.

Please find a way to fix it here and for all other numbered references in the article.

![image](https://github.com/user-attachments/assets/f8f9eb06-c417-4601-a2d3-e79a107b8725)